Terminal UI component

vinayaksh42 opened the pull request:
Fixes #188, Fixes #187, Fixes #165, Fixes #202, Fixes #205, and Fixes #207

[Preview link](https://saving-satoshi-git-fork-vinayaksh42-issue187-savingsatoshi.vercel.app/en/chapters/chapter-1/genesis-2)

changes:
1.  Disabled spell check and autocomplete for the terminal input area
2. paste now works in safari (iPhone)
3. Added command history option (use up and down arrow keys to cycle through previous commands)

A few changes that can be added:
1. disabling input when the challenge is completed (not sure if we should implement this change)
2. #145 - the [figma mockup](https://www.figma.com/file/LqjK3Tpvd9KJ4buFArCJBQ/Saving-Satoshi?node-id=1604%3A21036&t=aJplYJ7JVCvKwVOt-0) however has a normal terminal caret

Current known bugs:
1. terminal interface height not spanning over the entire div
3. Smooth Scroll behavior when the div expands

benalleng commented:
tACK on Firefox mobile

I implemented the input simply losing focus on a completed challenge so that the keyboard disappears. At the moment once you complete a challenge it's hard to tell what to do next. Especially at this early stage.
![Screenshot_20230226-101710](https://user-images.githubusercontent.com/108441023/221419562-00d29582-0f6e-4bb1-afca-28adff16caf9.png)

ecurrencyhodler commented:
Hey Ben. Good catch. The success text is currently being revamped in https://github.com/saving-satoshi/saving-satoshi/pull/194. We can probably add another line that says "Click next to continue" which should make the user at least want to look for a "next" button. But let me tag Christoph on this issue.

Vinayak, I tested paste functionality and it works great. I'm going to ask @afterburn to review the fix for command history. Lastly as I was testing spell check, I found a weird bug where the letters "fi" get smashed together if they're typed next to each other. I tried other letter combos but couldn't replicate. It's possible it could affect other letter combinations. 

I'll create another issue so we can track it properly but could you take a look to see if you can also include that in this PR?


<img width="619" alt="Screenshot 2023-02-26 at 8 58 36 AM" src="https://user-images.githubusercontent.com/32662508/221425005-9f8365a4-afe1-4cc5-8ca6-3c860a44415c.png">

vinayaksh42 commented:
the `fi` bug seems to be really weird. I tested it out on the [live website ](https://savingsatoshi.com/en/chapters/chapter-1/genesis-2) and the same behavior occurs. I think it might be due to the font style. Found [this](https://github.com/thespacedojo/crater-nova/issues/6) on the internet about it.

ecurrencyhodler commented:
Based on that convo, it sounds like there are two fixes:

1. disable ligatures-> font-feature-settings: "liga" 0;
2. add slight letter spacing

I'm feeling like the first option might be best. WDYT?

vinayaksh42 commented:
> Based on that convo, it sounds like there are two fixes:
> 
> 1. disable ligatures-> font-feature-settings: "liga" 0;
> 2. add slight letter spacing
> 
> I'm feeling like the first option might be best. WDYT?

yes, the first option seems to be the best one. I have implemented it.

ecurrencyhodler commented:
Sick. Thanks for taking on https://github.com/saving-satoshi/saving-satoshi/issues/202 as well.

afterburn commented on ui/lesson/Terminal.tsx:
```diff
@@ -21,27 +23,23 @@ export default function Terminal({ success, lines, next, onChange }) {
+                  'mt-2.5': index != 0,
+                })}
               >
                 <TerminalOutput>{`${line.value}`}</TerminalOutput>
```
I noticed `TerminalOutput` component is a single line of jsx, think it makes sense to just put that code here instead of having a dedicated component for it.

afterburn replied:
gh pr checkout 201

afterburn commented on ui/lesson/Terminal.tsx:
```diff
@@ -21,27 +23,23 @@ export default function Terminal({ success, lines, next, onChange }) {
-            colorMode={ColorMode.Light}
-            onInput={onChange}
-          >
+          <TerminalInterface onInput={onChange}>
```
I think there are a little too much Terminal-related components. I feel the logic that now resides in `TerminalInterface` could just be added to this component (`ui/lesson/Terminal.tsx`).

afterburn reviewed (changes requested):
Hey @vinayaksh42, great initial setup for a better terminal! I requested some small changes, but lgtm once you resolve them.

afterburn commented:
Actually there is another issue where if you lose focus, and then click on the input again you can end up in the following situation when pressing up/down arrow keys.

![saving-satoshi-terminal-bug](https://user-images.githubusercontent.com/10757768/221592035-a5162600-8a2f-48dc-bd59-b01cbd849614.gif)

benalleng reviewed (changes requested):
The fix for the "liga" font feature doesn't seem to be present in the Input field specifically
Perhaps explicitly specifying the font-family in the Input would solve this.

Also with large single words like hex values there isn't any word break so there are long horizontal scrollbars added to the terminal

![image](https://user-images.githubusercontent.com/108441023/221685821-d34d972b-c1f0-44ea-8ab8-548db424096e.png)

gadabra commented:
If `react-terminal-ui` package is not being used, you can remove it from the project.

afterburn commented:
> If `react-terminal-ui` package is not being used, you can remove it from the project.

I think there are a couple other packages we can also remove, feel free to address those in this PR as well @vinayaksh42

afterburn commented:
@vinayaksh42 I think the plan for the caret is to just use the normal one so maybe you can also tackle that in this PR.

vinayaksh42 commented:
I have removed `react-terminal-ui`, `date-fns`, and `react-individual-character-input-boxes` as they are not being used anymore. I have also added the `TerminalOutput` and `TerminalInterface` to the `Terminal` Component.

vinayaksh42 commented:
I have fixed the following issues:
- Lose focus upon challenge completion.
- the state of the challenge remains as completed upon getting the correct answer.

vinayaksh42 commented:
> Actually there is another issue where if you lose focus, and then click on the input again you can end up in the following situation when pressing up/down arrow keys.
> 
> ![saving-satoshi-terminal-bug](https://user-images.githubusercontent.com/10757768/221592035-a5162600-8a2f-48dc-bd59-b01cbd849614.gif)

@afterburn I am unable to reproduce this bug, can you tell me about the browser you are using?

ecurrencyhodler commented:
Amazing. 

- Confirmed that losing focus resolves https://github.com/saving-satoshi/saving-satoshi/issues/207.
- Confirmed that the state of the challenge remains as completed upon getting the correct answer resolves https://github.com/saving-satoshi/saving-satoshi/issues/205
- Confirmed link works and points to correct article.
![IMG_1526](https://user-images.githubusercontent.com/32662508/222282394-53edd25c-8e3d-4248-940e-6ee7fd762129.PNG)

ecurrencyhodler commented:
@vinayaksh42 I can confirm kevin's bug. If I click anywhere inside the console outside of the line where the carrot rests and start typing, nothing is entered. This occurs in chrome, firefox, and safari.

Also the carrot disappears on safari and firefox but is preserved in Chrome when I click outside the line. 

In Safari, the screen even scrolls when I press up or down.

This is pretty difficult to screen capture. If you can't reproduce I can hop on a call to show you.

ecurrencyhodler commented:
Okay the bug I found is different. Vinayak is going to address it in this PR.

@afterburn I am also unable to reproduce the error. Vinayak and I hopped on the call to go through it together in Safari/firefox/chrome.

afterburn commented:
Okay I installed another version of Chrome and the bug would not occur so it might be some weird setting or even an extension that causing the bug I reported. Feel free to ignore it.

benalleng reviewed (approved):
tACK for  issues #145, #187, #165, #202, #205, and #207
> 1. terminal interface height not spanning over the entire div

This bug has also been resolved

ecurrencyhodler commented:
@afterburn looks like vinayak addressed your requests. Would you mind doing a review then we can merge?

Thanks Ben for the tACK.
